### ISSUE TITLE
Comanche Maximum Overkill crashes DOSEMU

### ISSUE THREAD

**Issue by jschwartzenberg** [dosemu2/dosemu2#1703]
**Describe the bug**
Starting the game results in:
```
ERROR: general protection at 0x7f4c921038b8: 2e
```

**To Reproduce**
Run the game.

**Attach program/game binaries or provide an URL**
[COMANCHE.tar.gz](https://github.com/dosemu2/dosemu2/files/8985323/COMANCHE.tar.gz)

**Attach the log**
[boot.log](https://github.com/dosemu2/dosemu2/files/8985324/boot.log)

**A regression?**
unknown

**Comment by stsp** [comment]
`  000066b5: 0f22c0            05a7:0c45 mov  CR0,eax`
This game seems to
require real mode (to
enter prot mode by hands),
so it won't work presumably
also under win95.
Please check and if so -
lets close.

**Comment by volkertb** [comment]
Indeed, the initial version of _Comanche: Maximum Overkill_ was notorious in that it would switch to protected mode (or perhaps even 386 flat mode?) directly and would therefore refuse to run in VM86 mode. That meant that it could not be run under Windows, or in DOS when an EMM was loaded.

@stsp Wouldn't it be possible for dosemu2 to run such software, since VM86 mode wouldn't be strictly necessary to provide EMS memory to real-mode applications when leveraging hardware-assisted virtualization? I mean in the same way as how physical EMS boards back in the day would provide expanded memory without having the CPU run in VM86 mode. Couldn't dosemu2 just virtualize a "real" real mode environment that could be switched to protected mode or any other 386 mode by the application? Or does this perhaps have something to do with the fact that dosemu2 always provides DPMI, without any possibility to make that optional?

@jschwartzenberg The good news is that NovaLogic later made a patch available that made the game compatible with all memory modes, allowing it to run under Windows, as well as "in DOS without a boot disk". Presumably it accomplished this by patching a proper DOS extender into the game. There's a good chance that this patch will make the game run under dosemu2 as well. I've attached the patch below. Even though I uploaded it to [VirusTotal](https://www.virustotal.com) and no malware was detected there, any use is of course completely at your own risk.

[Comanche-Maximum-Overkill_Patch_DOS_EN.zip](https://github.com/dosemu2/dosemu2/files/9087638/Comanche-Maximum-Overkill_Patch_DOS_EN.zip)

I downloaded this patch from https://www.myabandonware.com/game/comanche-maximum-overkill-1sy (under "Game Extras")

[Ultima VII and Serpent Isle](https://nerdlypleasures.blogspot.com/2013/06/ultima-vii-on-real-hardware.html) are other examples, and (fan-made) [patches are available for those games as well](https://web.archive.org/web/20070124234755fw_/http://members.iinet.net.au/~rsd/five.html#XP). I don't know of any other games that have this problem, although I'm sure there are more. Thankfully, such problematic games are not common. The vast majority of DOS games are either real-mode games or use a DPMI-compatible DOS extender.

**Comment by stsp** [comment]
> VM86 mode wouldn't be strictly necessary to provide EMS

EMS has nothing to do with that indeed.

> Couldn't dosemu2 just virtualize a "real" real mode environment that could be switched to protected
> mode or any other 386 mode by the application?

That would be qemu.

> Or does this perhaps have something to do with the fact that dosemu2 always provides DPMI,
> without any possibility to make that optional?

This has something to do with
the fact that dosemu is not qemu. :)
Its a simple emulator for user-mode
code only.

> vast majority of DOS games are either real-mode games or use a DPMI-compatible DOS extender.

With the notable exception of
the pharlap's run286-based games,
and Origin's jemm API.
There are perhaps ~10 such games,
but they are "good ones", with a big
fan bases.

**Comment by stsp** [comment]
> EMS has nothing to do with that indeed.

Think of emm286.exe - it doesn't
need real "real" mode. It just works.

**Comment by jschwartzenberg** [comment]
Indeed the patch makes it run!
There are quite a few glitches though.
![Screenshot_20220713_214520](https://user-images.githubusercontent.com/877112/178818682-073a0291-2c25-4aa7-ac7d-cea30dc7dc01.png)

@stsp you prefer to move to another ticket or shall we just continue in this one? The patch part could simply be added to the documentation at least.

**Comment by Baron-von-Riedesel** [comment]
Just tried the patched variant of the game with the current DosEmu github: sound works, but gfx has severe problems.
In boot.log, there are these strange lines:
```
CONF: reserving 16Kb at 0xE8000 for 'E' (EMS page frame)
CONF: reserving 16Kb at 0xEC000 for 'E' (EMS page frame)
Registering HWRAM, type=S base=0x40014000 size=0x10000
CONF: reserving 64Kb at 0x40014000 for 'S' ((null))
Registering HWRAM, type=S base=0x40024000 size=0x10000
CONF: reserving 64Kb at 0x40024000 for 'S' ((null))
Registering HWRAM, type=S base=0x40014000 size=0x10000
CONF: reserving 64Kb at 0x40014000 for 'S' ((null))
Registering HWRAM, type=S base=0x40024000 size=0x10000
CONF: reserving 64Kb at 0x40024000 for 'S' ((null))
eVGAEmuFault: found vga_page 7 map 0
[snip (deleted a few dozens similar lines)]
eVGAEmuFault: found vga_page 7 map 0
eVGAEmuFault: found vga_page 8 map 0
[snip (deleted a few dozens similar lines)]
eVGAEmuFault: found vga_page 8 map 0
eVGAEmuFault: found vga_page 2 map 0
```

**Comment by stsp** [comment]
Confirmed, complete garbage.
Not sure if I'll get to this soon,
maybe Bart can look? :)

**Comment by bartoldeman** [comment]
Sure I can have a look again at vgaemu, now the PC speaker work seems to be done.
